### PR TITLE
[Integration Test] Itemized stateful storage, e2e test [Merge After #1036 ]

### LINKED ISSUE #1021: [Private Graph] integration test for itemized storage
TBD

Demo script:
as a user I would like to write graph data to itemized storage that are my keys from the provider I wanted to connect.

- posting extrinsic with one of the action in the array to either read or write(right now just keys)
- edit extrinsic with one of the action in the array to either read or write(right now just keys)

### LINKED ISSUE #1022: [Private Graph] integration test for paginated storage
TBD

Demo script:
as a user I would like to write graph data to paginated storage that ______

- 
-

### PR DESCRIPTION
**Pull request by saraswatpuneet**
# Goal
The goal of this PR is  to add integration tests for itemized storage types

Closes #1021
Closes #1022

### CODE REVIEW

File: integration-tests/stateful-pallet-storage/handleItemized.test.ts [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        if (providerEvent && ExtrinsicHelper.api.events.msa.ProviderCreated.is(providerEvent)) {
+            providerId = providerEvent.data.providerId;
+        }
+        assert.notEqual(providerId, undefined, "setup should populate providerId");
+
+        // Create a schema for Itemized PayloadLocation
+        const createSchema = ExtrinsicHelper.createSchema(providerKeys, AVRO_CHAT_MESSAGE, "AvroBinary", "Itemized");
+        const [event] = await createSchema.fundAndSend();
+        if (event && createSchema.api.events.schemas.SchemaCreated.is(event)) {
+            [, schemaId] = event.data;
```
Comment (JoeCap08055) [inline integration-tests/stateful-pallet-storage/handleItemized.test.ts:35]: This has been fixed now with #961 so you can do:
`schemaId = event.data.schemaId;`

File: integration-tests/stateful-pallet-storage/handleItemized.test.ts [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+            }
+
+            let add_actions = [add_action, update_action];
+            let itemized_add_result_1 = ExtrinsicHelper.applyItemActions(providerKeys, schemaId, msa_id, add_actions);
+            await itemized_add_result_1.fundOperation();
+            const [pageUpdateEvent1, chainEvents ] = await itemized_add_result_1.signAndSend();
+            assert.notEqual(chainEvents["system.ExtrinsicSuccess"], undefined, "should have returned an ExtrinsicSuccess event");
+            assert.notEqual(chainEvents["transactionPayment.TransactionFeePaid"], undefined, "should have returned a TransactionFeePaid event");
+            assert.notEqual(pageUpdateEvent1, undefined, "should have returned a PalletStatefulStorageItemizedActionApplied event");
+            // Delete action
```
Comment (JoeCap08055) [inline integration-tests/stateful-pallet-storage/handleItemized.test.ts:105]: Maybe break this next section out into a separate test block? I think single-/limited-purpose tests are good style, and easier to quickly see what failed.

File: integration-tests/stateful-pallet-storage/handleItemized.test.ts [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+            const add_action = {
+                "Add" : payload_1
+            }
+            let add_actions = [add_action];    
+            let itemized_add_result_1 = ExtrinsicHelper.applyItemActions(providerKeys, schemaId_unsupported, msa_id, add_actions);
+            await itemized_add_result_1.fundOperation();
+            await assert.rejects(async () => {
+                await itemized_add_result_1.signAndSend();
+            }
+            , (err) => {
```
Comment (JoeCap08055) [inline integration-tests/stateful-pallet-storage/handleItemized.test.ts:152]: You can put the actual error enum expected here instead of an error handler

**Review by JoeCap08055 — COMMENTED** [review]
Seems like a lot of duplication of setup between paginated & itemized, by having them split into 2 files. Why not combine them, do common provider & MSA setup for both, and split the tests into 2 `describe()` blocks?

**Comment by saraswatpuneet** [comment]
📗 Stateful Pallet Storage
    Itemized Storage Tests 😊/😥
      ✔ ✅ should be able to call applyItemizedAction and apply actions (57ms)
      ✔ 🛑 should fail to call applyItemizedAction with invalid schemaId (67ms)
      ✔ 🛑 should fail to call applyItemizedAction with invalid schema location (52ms)
      ✔ 🛑 should fail to call applyItemizedAction with for un-delegated attempts (56ms)
    Itemized Storage Remove Action Tests
      ✔ ✅ should be able to call applyItemizedAction and apply remove actions (117ms)
      ✔ 🛑 should fail to call remove action with invalid schemaId (61ms)
      ✔ 🛑 should fail to call remove action with invalid schema location (62ms)
      ✔ 🛑 should fail to call remove action with invalid msa_id (52ms)

  📗 Stateful Pallet Storage
    Paginated Storage Upsert/Remove Tests 😊/😥
      ✔ ✅ should be able to call upsert page and add a page and remove a page via id (166ms)
      ✔ 🛑 should fail to call upsert page with invalid schemaId (54ms)
      ✔ 🛑 should fail to call upsert page with invalid schema location (65ms)
      ✔ 🛑 should fail to call upsert page with for un-delegated attempts (51ms)
    Paginated Storage Removal Negative Tests 😊/😥
      ✔ 🛑 should fail to call remove page with invalid schemaId (56ms)
      ✔ 🛑 should fail to call remove page with invalid schema location (51ms)
      ✔ 🛑 should fail to call remove page with for un-delegated attempts (57ms)

File: integration-tests/stateful-pallet-storage/handleItemized.test.ts (hunk 3 above)
Comment (saraswatpuneet) [reply integration-tests/stateful-pallet-storage/handleItemized.test.ts:108]: Done :)

File: integration-tests/scaffolding/helpers.ts [hunk 4]
```diff
@@ -73,3 +74,41 @@ export function log(...args: any[]) {
...
+        schemaIds: [schemaId],
+    });
+    const addProviderData = ExtrinsicHelper.api.registry.createType("PalletMsaAddProvider", payload);
+
+    const grantDelegationOp = ExtrinsicHelper.grantDelegation(keys, providerKeys, signPayloadSr25519(keys, addProviderData), payload);
+    await grantDelegationOp.fundOperation();
+    await grantDelegationOp.signAndSend();
+
+    return [keys, delegator_msa_id];
+}
```
Comment (saraswatpuneet) [inline integration-tests/scaffolding/helpers.ts:114]: Some helper functions to create Provider and create Delegator with delegation activated

File: integration-tests/stateful-pallet-storage/handleItemized.test.ts [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+            let add_actions = [add_action, update_action];
+            let itemized_add_result_1 = ExtrinsicHelper.applyItemActions(providerKeys, schemaId, msa_id, add_actions);
+            await itemized_add_result_1.fundOperation();
+            const [pageUpdateEvent1, chainEvents ] = await itemized_add_result_1.signAndSend();
+            assert.notEqual(chainEvents["system.ExtrinsicSuccess"], undefined, "should have returned an ExtrinsicSuccess event");
+            assert.notEqual(chainEvents["transactionPayment.TransactionFeePaid"], undefined, "should have returned a TransactionFeePaid event");
+            assert.notEqual(pageUpdateEvent1, undefined, "should have returned a PalletStatefulStorageItemizedActionApplied event");
+        }).timeout(10000);
+
+        it("🛑 should fail to call applyItemizedAction with invalid schemaId", async function () {
```
Comment (aramikm) [inline integration-tests/stateful-pallet-storage/handleItemized.test.ts:75]: Do we really need these simple failure test cases that should be all covered by unit tests?  My understanding was that we are going to use integration tests for more complicated use-cases.

File: pallets/stateful-storage/src/types.rs [hunk 6]
```diff
@@ -95,19 +95,23 @@ impl<PageDataSize: Get<u32>> Page<PageDataSize> {
 				},
 			}
 		}
+		log::info!("add_buffer len: {:?}", add_buffer.len());
```
Comment (enddynayn) [inline pallets/stateful-storage/src/types.rs:98]: are these needed?

Comment (saraswatpuneet) [reply pallets/stateful-storage/src/types.rs]: Joe probably cleaned up in #1036 but will check

**Comment by saraswatpuneet** [comment]
**Added test for Stateful Storage RPCs**


  📗 Stateful Pallet Storage
    Itemized Storage Tests 😊/😥
      ✔ ✅ should be able to call applyItemizedAction and apply actions (72ms)
      ✔ 🛑 should fail to call applyItemizedAction with invalid schemaId (72ms)
      ✔ 🛑 should fail to call applyItemizedAction with invalid schema location (83ms)
      ✔ 🛑 should fail to call applyItemizedAction with for un-delegated attempts (68ms)
    Itemized Storage Remove Action Tests
      ✔ ✅ should be able to call applyItemizedAction and apply remove actions (188ms)
      ✔ 🛑 should fail to call remove action with invalid schemaId (75ms)
      ✔ 🛑 should fail to call remove action with invalid schema location (58ms)
      ✔ 🛑 should fail to call remove action with invalid msa_id (66ms)
    Itemized Storage RPC Tests
      ✔ ✅ should be able to call get_itemized_storages

  📗 Stateful Pallet Storage
    Paginated Storage Upsert/Remove Tests 😊/😥
      ✔ ✅ should be able to call upsert page and add a page and remove a page via id (182ms)
      ✔ 🛑 should fail to call upsert page with invalid schemaId (61ms)
      ✔ 🛑 should fail to call upsert page with invalid schema location (58ms)
      ✔ 🛑 should fail to call upsert page with for un-delegated attempts (110ms)
    Paginated Storage Removal Negative Tests 😊/😥
      ✔ 🛑 should fail to call remove page with invalid schemaId (55ms)
      ✔ 🛑 should fail to call remove page with invalid schema location (61ms)
      ✔ 🛑 should fail to call remove page with for un-delegated attempts (65ms)
    Paginated Storage RPC Tests
      ✔ ✅ should be able to call get_paginated_storages

**Review by aramikm — APPROVED** [review]
Ship it!

**Pull request merged**